Add docs for fast cutback in v24.1 physical cluster replication (#18506)
cockroachdb/docs: src/current/v24.1/cutover-replication.md

Initiating a cutover is a manual process that makes the standby cluster ready to accept SQL connections. However, the cutover process does **not** automatically redirect traffic to the standby cluster. Once the cutover is complete, you must redirect application traffic to the standby (new) cluster. If you do not manually redirect traffic, writes to the primary (original) cluster may be lost.
{{site.data.alerts.end}}

> After a cutover, you may want to _cut back_ to the original primary cluster (or a different cluster). That is, set up the original primary cluster to once again accept application traffic. For more details, refer to [Cut back to the primary cluster](#cut-back-to-the-primary-cluster).

Reviewer: ```suggestion
After a cutover, you may want to _cut back_ to the original primary cluster (or a different cluster) to set up the original primary cluster to once again accept application traffic. For more details, refer to [Cut back to the primary cluster](#cut-back-to-the-primary-cluster).
```

## Cut back to the primary cluster

> After cutting over to the standby cluster, you may need to move back to the original primary cluster to serve your application.

Reviewer: I'd lean into either "cut back" or "move back", but it's a little confusing to see both.

{% include {{ page.version.version }}/physical-replication/fast-cutback-syntax.md %}

{{site.data.alerts.callout_info}}
To move back to a different cluster, follow the physical cluster replication [setup]({% link {{ page.version.version }}/set-up-physical-cluster-replication.md %}).
{{site.data.alerts.end}}

### Example

> The following is a step-by-step example of cutting back to the original primary cluster. This example assumes that a cutover from a primary to standby cluster has occurred, and the promoted standby has been serving traffic.

Reviewer: ```suggestion
This section illustrates the steps to cut back to the original primary cluster after cutting over to the standby cluster when the promoted standby is currently serving traffic.
```

> - **Cluster A** = original primary cluster

Reviewer: This may be easier to follow with more descriptive names, like `original_standby_cluster`, `original_primary_cluster`?

- **Cluster B** = original standby cluster

> 1. **Cluster B** is now serving application traffic after the [cutover](#step-2-complete-the-cutover).

Reviewer: ```suggestion
1. **Cluster B** is serving application traffic after the [cutover](#step-2-complete-the-cutover).
```

> 1. To begin the cutback to the original primary cluster (**cluster A**), it is necessary that the virtual cluster (e.g., `vc_a`) is no longer serving traffic. Connect to the system virtual on **cluster A**:

Reviewer: ```suggestion
1. To begin the cutback to the original primary cluster (**cluster A**), the virtual cluster (e.g., `vc_a`) must first stop accepting connections. Connect to the system virtual on **cluster A**:
```

Where did `vc_a` come from?

Author: I got rid of that and added `cluster_a` instead, no explanation needed for that, so better!

    {% include_cached copy-clipboard.html %}
    ~~~ shell
    cockroach sql --url \
    "postgresql://{user}@{node IP or hostname cluster A}:26257?options=-ccluster=system&sslmode=verify-full" \
    --certs-dir "certs"
    ~~~

> 1. From the system virtual cluster on **cluster A**, ensure that service to the virtual cluster has stopped:

Reviewer: ```suggestion
1. From the system virtual cluster on **cluster A**, stop the service:
```

    {% include_cached copy-clipboard.html %}
    ~~~ sql
    ALTER VIRTUAL CLUSTER vc_a STOP SERVICE;
    ~~~

> 1. Open another terminal window and connect to the system virtual cluster for **cluster B**:

Reviewer: ```suggestion
1. Open another terminal window and connect to the system virtual cluster for **Cluster B**:
```
Maybe check for consistency on the cluster names.

    {% include_cached copy-clipboard.html %}
    ~~~ shell
    cockroach sql --url \
    "postgresql://{user}@{node IP or hostname cluster B}:26257?options=-ccluster=system&sslmode=verify-full" \
    --certs-dir "certs"
    ~~~

1. From the system virtual cluster on **cluster B**, enable rangefeeds:

    {% include_cached copy-clipboard.html %}
    ~~~ sql
    SET CLUSTER SETTING kv.rangefeed.enabled = 'true';
    ~~~

1. From the system virtual cluster on **cluster A**, start the replication from cluster B to cluster A:

    {% include_cached copy-clipboard.html %}
    ~~~ sql
    ALTER VIRTUAL CLUSTER {vc_a} START REPLICATION OF {vc_b} ON 'postgresql://{user}@{ node IP or hostname cluster B}:26257?options=-ccluster=system&sslmode=verify-full&sslrootcert=certs/{standby cert}.crt';
    ~~~

>     This will rewind the virtual cluster on **A** back to the time at which the same virtual cluster on **B** diverged from it. **Cluster A** will check with **cluster B** to confirm that its virtual cluster was replicated from **cluster A** as part of the original [physical cluster replication stream]({% link {{ page.version.version }}/set-up-physical-cluster-replication.md %}).

Reviewer: If you take the above suggestion about "rewind", this should be updated too.

    {{site.data.alerts.callout_success}}
    For details on connection strings, refer to the [Connection reference]({% link {{ page.version.version }}/set-up-physical-cluster-replication.md %}#connection-reference).
    {{site.data.alerts.end}}

1. Check the status of the virtual cluster on **A**:

    {% include_cached copy-clipboard.html %}
    ~~~ sql
    SHOW VIRTUAL CLUSTER {vc_a};
    ~~~

    {% include_cached copy-clipboard.html %}
    ~~~
     id |  name  |     data_state     | service_mode
    ----+--------+--------------------+---------------
      1 | system | ready              | shared
      3 | {vc_a} | replicating        | none
      4 | test   | replicating        | none
      (2 rows)
    ~~~

> 1. Run the following from **cluster A** to start the cutover:

Reviewer: ```suggestion
1. From **cluster A**, start the cutover:
```

    {% include_cached copy-clipboard.html %}
    ~~~ sql
    ALTER VIRTUAL CLUSTER {vc_a} COMPLETE REPLICATION TO LATEST;
    ~~~

    The `cutover_time` is the timestamp at which the replicated data is consistent. The cluster will revert any data above this timestamp:

    ~~~
               cutover_time
    ----------------------------------
      1714497890000000000.0000000000
    (1 row)
    ~~~

> 1. From **cluster A**, bring the virtual cluster online with:

Reviewer: ```suggestion
1. From **cluster A**, bring the virtual cluster online:
```

    {% include_cached copy-clipboard.html %}
    ~~~ sql
    ALTER VIRTUAL CLUSTER {vc_a} START SERVICE SHARED;
    ~~~

> 1. To make **cluster A's** virtual cluster the default for connection strings, set the following [cluster setting]({% link {{ page.version.version }}/cluster-settings.md %}):

Reviewer: Maybe link to https://www.cockroachlabs.com/docs/stable/work-with-virtual-clusters#sql-clients?

    {% include_cached copy-clipboard.html %}
    ~~~ sql
    SET CLUSTER SETTING server.controller.default_target_cluster='{vc_a}';
    ~~~

> At this point, **cluster A** is once again the primary and **cluster B** is once again the standby. The clusters are entirely independent. You will need to use your own network load balancers, DNS servers, or other network configuration to direct application traffic to the primary (**cluster A**). To enable physical cluster replication again, from the primary to the standby (or a completely different cluster), refer to [Set Up Physical Cluster Replication]({% link {{ page.version.version }}/set-up-physical-cluster-replication.md %}).

Reviewer: ```suggestion
At this point, **cluster A** is once again the primary and **cluster B** is once again the standby. The clusters are entirely independent. To direct application traffic to the primary (**Cluster A**), you will need to use your own network load balancers, DNS servers, or other network configuration to direct application traffic to the primary (**cluster A**). To enable physical cluster replication again, from the primary to the standby (or a completely different cluster), refer to [Set Up Physical Cluster Replication]({% link {{ page.version.version }}/set-up-physical-cluster-replication.md %}).
```
If Cluster B is once again the standby after the above, then I think the last sentence isn't needed. WDYT?

Author: Why? It is just pointing users back to the page to set up physical cluster replication, unless I am missing something.

    {% include_cached copy-clipboard.html %}
    ~~~ sql
    SET CLUSTER SETTING server.controller.default_target_cluster='{vc_a}';
    ~~~

At this point, **cluster A** is once again the primary and **cluster B** is once again the standby. The clusters are entirely independent. You will need to use your own network load balancers, DNS servers, or other network configuration to direct application traffic to the primary (**cluster A**). To enable physical cluster replication again, from the primary to the standby (or a completely different cluster), refer to [Set Up Physical Cluster Replication]({% link {{ page.version.version }}/set-up-physical-cluster-replication.md %}).

### Cut back to a different cluster

To create a completely new virtual cluster in the original primary cluster, or another physical cluster, use the [`CREATE VIRTUAL CLUSTER`]({% link {{ page.version.version }}/create-virtual-cluster.md %}) syntax:

{% include_cached copy-clipboard.html %}
~~~ sql
CREATE VIRTUAL CLUSTER new_vc FROM REPLICATION OF promoted_standby_vc ON connection_string_standby;

> ~~~

Author: I don't know if we really need/want this section given that it is just starting a new replication stream. Open to removing and maybe just adding a note.

Reviewer 2: yeah, I don't really see what this is adding

### Example

This section illustrates the steps to cut back to the original primary cluster from the promoted standby cluster that is currently serving traffic.

- **Cluster A** = original primary cluster
- **Cluster B** = original standby cluster

> 1. **Cluster B** is serving application traffic after the [cutover](#step-2-complete-the-cutover).

Reviewer: Nit: It doesn't seem quite right that this is a numbered step. Maybe it should be a paragraph after the list.
